// @prosemirror-adapter/core/lib/index.es.js
var D = (o, t, s) => {
  if (!t.has(o))
    throw TypeError("Cannot " + s);
};
var f = (o, t, s) => {
  if (t.has(o))
    throw TypeError("Cannot add the same private member more than once");
  t instanceof WeakSet ? t.add(o) : t.set(o, s);
};
var l = (o, t, s) => (D(o, t, "access private method"), s);
var d, m;
class O {
  constructor({ node: t, view: s, getPos: n, decorations: a, innerDecorations: w, options: h }) {
    f(this, d);
    this.selected = !1, this.selectNode = () => {
      var e, i;
      this.selected = !0, (i = (e = this.options).selectNode) == null || i.call(e);
    }, this.deselectNode = () => {
      var e, i;
      this.selected = !1, (i = (e = this.options).deselectNode) == null || i.call(e);
    }, this.shouldUpdate = (e) => !(e.type !== this.node.type || e.sameMarkup(this.node) && e.content.eq(this.node.content)), this.update = (e, i, r) => {
      var M, g;
      const c = this.options.update;
      let u;
      return c && (u = c(e, i, r)), typeof u != "boolean" && (u = this.shouldUpdate(e)), this.node = e, this.decorations = i, this.innerDecorations = r, u && ((g = (M = this.options).onUpdate) == null || g.call(M)), u;
    }, this.shouldIgnoreMutation = (e) => !this.dom || !this.contentDOM || this.node.isLeaf || this.node.isAtom ? !0 : e.type === "selection" ? !1 : this.contentDOM === e.target && e.type === "attributes" ? !0 : !this.contentDOM.contains(e.target), this.ignoreMutation = (e) => {
      if (!this.dom || !this.contentDOM)
        return !0;
      let i;
      const r = this.options.ignoreMutation;
      return r && (i = r(e)), typeof i != "boolean" && (i = this.shouldIgnoreMutation(e)), i;
    }, this.destroy = () => {
      var e, i, r;
      (i = (e = this.options).destroy) == null || i.call(e), this.dom.remove(), (r = this.contentDOM) == null || r.remove();
    }, this.setAttrs = (e) => {
      const { dispatch: i, state: r } = this.view, c = this.getPos();
      if (typeof c == "number")
        return i(
          r.tr.setNodeMarkup(c, void 0, {
            ...this.node.attrs,
            ...e
          })
        );
    }, this.node = t, this.view = s, this.getPos = n, this.decorations = a, this.innerDecorations = w, this.options = h, this.dom = this.createDOM(h.as), this.contentDOM = t.isLeaf ? null : this.createContentDOM(h.contentAs), this.dom.setAttribute("data-node-view-root", "true"), this.contentDOM && (this.contentDOM.setAttribute("data-node-view-content", "true"), this.contentDOM.style.whiteSpace = "inherit"), this.setSelection = h.setSelection, this.stopEvent = h.stopEvent;
  }
  createDOM(t) {
    return l(this, d, m).call(this, t);
  }
  createContentDOM(t) {
    return l(this, d, m).call(this, t);
  }
  get component() {
    return this.options.component;
  }
}
d = new WeakSet(), m = function(t) {
  const { node: s } = this;
  return t == null ? document.createElement(s.isInline ? "span" : "div") : t instanceof HTMLElement ? t : t instanceof Function ? t(s) : document.createElement(t);
};
class E {
  constructor(t) {
    this.view = t.view, this.options = t.options;
  }
  get component() {
    return this.options.component;
  }
  get root() {
    var s, n;
    let t = (n = (s = this.options).root) == null ? void 0 : n.call(s, this.view.dom);
    return t || (t = this.view.dom.parentElement ?? document.body), t;
  }
  update(t, s) {
    var n, a;
    this.view = t, this.prevState = s, (a = (n = this.options).update) == null || a.call(n, t, s);
  }
  destroy() {
    var t, s;
    (s = (t = this.options).destroy) == null || s.call(t);
  }
}
var p, v;
class b {
  constructor({ pos: t, spec: s, options: n }) {
    f(this, p);
    this.pos = t, this.options = n, this.spec = s, this.dom = l(this, p, v).call(this, n.as), this.dom.setAttribute("data-widget-view-root", "true");
  }
  bind(t, s) {
    this.view = t, this.getPos = s;
  }
  get component() {
    return this.options.component;
  }
}
p = new WeakSet(), v = function(t) {
  return t instanceof HTMLElement ? t : document.createElement(t);
};
export {
  O as CoreNodeView,
  E as CorePluginView,
  b as CoreWidgetView
};
